### ISSUE TITLE
Support for SSH check mode in ACLs

### ISSUE THREAD

**Issue by almereyda** [juanfont/headscale#1850]
## Why

Tailscale upstream supports SSH check mode.

We would like to use it with Headscale, too.

## Description

When defining an Tailscale SSH ACL policy with the `action` set to `check`, an additional authentication against the OIDC endpoint is required, which grants access within an optional `checkPeriod`, defaulting to 12 hours and allowing to be set to `always`.

## References

* [Configure Tailscale SSH with check · Tailscale SSH · Tailscale Docs](https://tailscale.com/kb/1193/tailscale-ssh#configure-tailscale-ssh-with-check-mode)

This is related to, but not identical to:

- #1303, where a user tried to use `check` mode
- #1623, where upstream reports about the availability of a web check mode in the Tailscale client, currently restricted to the upstream control plane.

**Comment by almereyda** [comment]
No stale activity.

**Comment by dparv** [comment]
+1

**Comment by Codelica** [comment]
I took a look at this also, as we'd _gladly_ pay a bounty for SSH check mode support.  (Sadly we have no Go devs).  

Looking at the policy code for both V1 and V2, they both seem to use the `checkPeriod` to set `SessionDuration` in the response:

V1:
https://github.com/juanfont/headscale/blob/43943aeee9134cf6a76a380ebb1bc3ac7803d830/hscontrol/policy/v1/acls.go#L373
V2:
https://github.com/juanfont/headscale/blob/43943aeee9134cf6a76a380ebb1bc3ac7803d830/hscontrol/policy/v2/filter.go#L92

Which according to Tailscale code seems to be "how long the session can stay open before being forcefully terminated":

https://github.com/tailscale/tailscale/blob/fccba5a2f1a5e5dbde9e2fa57e33651b8fd047eb/tailcfg/tailcfg.go#L2692

So it seems this was implemented incorrectly.  

My guess is that a conditional combination of `Message` and `HoldAndDelegate` is what would be used to block/wait on verification if the user hasn't authenticated within the check period:

https://github.com/tailscale/tailscale/blob/fccba5a2f1a5e5dbde9e2fa57e33651b8fd047eb/tailcfg/tailcfg.go#L2700

I know the focus is currently on the policy v2 re-write beta, but perhaps it could be reviewed as a bug?  :) If not, maybe it's a candidate for the next version?

Thanks for reading...

**Comment by kradalby** [comment]
@Codelica I noticed as well that there was something funky with the implementation, but I tried to separate the 0.26 release to be the rewrite so the feature creep it. I'm going to release it now, and then we start a new cycle which I will add this as a bug that needs to be fixed.

> gladly pay a bounty for SSH check mode support

As people know, I am paid, so not needed, but donating to Juan/The project is always welcome to cover costs for anything from domain and our dream about putting up own build servers.

**Comment by Codelica** [comment]
Excellent!  Count us in for testing. 👍   Will get migrated to 0.26.

**Comment by Codelica** [comment]
I can definitely appreciate that (testing difficulty) with time, oidc and ssh involved.  It looks like `checkPeriod` has a minimum of 1 minute and maximum of 168 hours (1 week), along with an "always" option to always check.  But even a test scenario using a short duration, delays, etc could get complicated depending how far it's taken (re-auth declines/failures, etc). 

Personally I think it's a key feature for Tailscale SSH access though (especially for connections left up 24x7), and would love to see it working.   I can promise once it's functional we'll definitely be doing a "real world" test before deploying any new version to production.  If you'd consider adding a final RC release (or somehow indicating a final beta), we could definitely promise to test those before new version releases.  I realize that's not as good as an automated test that can be continually run, but hopefully better than nothing. If the logic is isolated and written defensively when possible (fail securely, etc), maybe that and hand testing is enough to start? :)

**Comment by kradalby** [comment]
I've been working on SSH check mode in https://github.com/juanfont/headscale/pull/3076, it now passes the initial integration tests and I would find it incredibly helpful if people could help me take it for a spin (and try to break it).

**Comment by kradalby** [comment]
This is now in the main branch, please help test!

**Comment by Codelica** [comment]
Thanks... I REALLY do want to test this out, but have some blockers this week.  I will definitely try next week though.  Thank you.

**Comment by kradalby** [comment]
Great! It will make it into a beta, but we have quite a lot more on the milestone before we release that.

**Comment by Codelica** [comment]
I don't mind waiting for a beta image (total work chaos at the moment), but don't want to hold things up if your looking for feedback now.  Any chance for a tagged dev build that includes this?  Docker Hub didn't seem to have any candidates. Thanks!

**Comment by lordwelch** [comment]
Both of my notes about differences are resolved now. I haven't run into any further issues

**Comment by kradalby** [comment]
A new beta is out: [v0.29.0-beta.1](https://github.com/juanfont/headscale/releases/tag/v0.29.0-beta.1).

This feature is included in the release.

If you discover problems with it, open a new issue with full debug information: headscale config, policy, headscale logs at trace level, plus `tailscale debug netmap` and `tailscale status --json` from an affected client. See the [debugging and troubleshooting guide](https://headscale.net/development/ref/debug/) for collection steps and follow the issue template.
